File: 2721-sum-of-distances/sum-of-distances.java
class Solution {
    public long[] distance(int[] nums) {
        Map<Integer, Integer> freq = new HashMap<>();
        Map<Integer, Long> sum = new HashMap<>();

        long[] res = new long[nums.length];

        for (int i = 0; i < nums.length; i++) {
            if (freq.containsKey(nums[i])) {
                int f = freq.get(nums[i]);
                long s = sum.get(nums[i]);

                res[i] += (long) i * f - s;

                freq.put(nums[i], f + 1);
                sum.put(nums[i], s + i);
            } else {
                freq.put(nums[i], 1);
                sum.put(nums[i], (long) i);
            }
        }

        freq.clear();
        sum.clear();

        for (int i = nums.length - 1; i >= 0; i--) {
            if (freq.containsKey(nums[i])) {
                int f = freq.get(nums[i]);
                long s = sum.get(nums[i]);

                res[i] += s - (long) i * f;

                freq.put(nums[i], f + 1);
                sum.put(nums[i], s + i);
            } else {
                freq.put(nums[i], 1);
                sum.put(nums[i], (long) i);
            }
        }

        return res;
    }
}

File: 2721-sum-of-distances/sum-of-distances.py
class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n  # Initialize result array with zeros

        # First pass: left to right
        # freq: stores count of each number encountered
        # sum_indices: stores sum of indices for each number encountered
        freq = {}
        sum_indices = {}

        for i in range(n):
            num = nums[i]
            if num in freq:
                count = freq[num]
                prev_sum = sum_indices[num]

                # Sum of distances to elements on the left:
                # (i - j1) + (i - j2) + ... = i * count - (j1 + j2 + ...)
                res[i] += i * count - prev_sum

                freq[num] += 1
                sum_indices[num] += i
            else:
                # First time seeing this number
                freq[num] = 1
                sum_indices[num] = i

        # Clear maps for the second pass
        freq.clear()
        sum_indices.clear()

        # Second pass: right to left
        for i in range(n - 1, -1, -1): # Iterate from n-1 down to 0
            num = nums[i]
            if num in freq:
                count = freq[num]
                next_sum = sum_indices[num]

                # Sum of distances to elements on the right:
                # (j1 - i) + (j2 - i) + ... = (j1 + j2 + ...) - i * count
                res[i] += next_sum - i * count

                freq[num] += 1
                sum_indices[num] += i
            else:
                # First time seeing this number from the right
                freq[num] = 1
                sum_indices[num] = i
                
        return res